campaignion_email_to_target/campaignion_email_to_target.module

```diff
@@ -153,35 +155,70 @@ function campaignion_email_to_target_token_info() {
       drupal_set_message(t('We are sorry! The service is temporary unavailable. The administrators have been informed. Please try again in a few minutes …'), 'error');
     }
   }
+
+  $info['tokens']['email-to-target']['messages'] = [
+    'name' => t('The messages sent to targets'),
+    'description' => t('The messages sent to E2T targets.'),
+  ];
+
   return $info;
 }
 
 /**
  * Implements hook_tokens().
  */
 function campaignion_email_to_target_tokens($type, array $tokens, array $data = [], array $options = []) {
-  if ($type != 'email-to-target' || empty($data['email-to-target'])) {
-    return;
+  $replacements = [];
+  if ($type != 'email-to-target') {
+    return $replacements;
   }
 
-  $my_data['contact'] = $data['email-to-target'];
-  foreach ($my_data['contact'] as $key => $sub_array) {
-    if (is_array($sub_array)) {
-      $my_data[$key] = $sub_array;
+  // Token for complete messages sent to target.
```

Comment:
I’d suggest splitting the token callback function. We could put them into 1 or 2 new classes.